### PR TITLE
Add IoPathSaver and unify the mode for Saver

### LINKED ISSUE #91: Use iopath in `SaverIterDataPipe`
## 🚀 Feature

`iopath` provides an API that can replace `open` and support saving to multiple destinations (including S3), we should probably reimplement `SaverIterDataPipe` to rely on `iopath` instead.

We could even then rename the functional API so that it's called `save` instead of `save_to_disk`.

### PR DESCRIPTION
**Pull request by ejguan**
Fixes #91

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#105 Add IoPathSaver and unify the mode for Saver**
* #104 Refactor IoPath FileLister and FileLoader

Differential Revision: [D32317606](https://our.internmc.facebook.com/intern/diff/D32317606)

### CODE REVIEW

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32317606).

**Comment by ejguan** [comment]
This is also BC-breaking change. But, no need to let domains to change their code because `Saver` is not used by them directly. The `on_disk_cache` is the main consumer of `Saver` and it would explicitly add mode to `Saver`

File: torchdata/datapipes/iter/util/saver.py [hunk 1]
```diff
@@ -15,19 +16,20 @@ class SaverIterDataPipe(IterDataPipe[str]):
...
+        filepath_fn: Function that takes in metadata nad returns the target path of the new file
     """
+
     def __init__(
         self,
-        source_datapipe: IterDataPipe[Tuple[Any, IOBase]],
-        mode: str = "wb",
-        filepath_fn: Callable = _default_filepath_fn,
+        source_datapipe: IterDataPipe[Tuple[Any, U]],
+        mode: str = "w",
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/saver.py:27]: BC breaking here.

**Review by NivekT — APPROVED** [review]
LGTM

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32317606).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32317606).

**Pull request closed without merging**